### ISSUE TITLE
`supports_value_expectations?` breaks `have` collection matcher for `should` one-liner syntax

### ISSUE THREAD

**Issue by amilligan** [rspec/rspec-collection_matchers#50]
Please, I beg of you, let's not rehash RSpec grammar wars.  The `is_expected.to` syntax is newer, but the `should` one-liner syntax is supported and should behave appropriately.

### Subject of the issue
As of the deprecation of implicit block expectation syntax, and introduction of the check for use of this syntax (version 3.11.0), the `have` matcher (from rspec-collection_matchers) no longer works properly with an explicit subject.  E.g.:

```ruby
  describe "#wibbles" do
    subject { foo }
    it { should have(3).wibbles } # => does not work
    it { is_expected.to have(3).wibbles } # => works
  end
```
For the case with the `should` syntax this results in:
`NoMethodError: undefined method `supports_value_expectations?' for #<Foo:0x00007ffdadd0e2a8>`

If you change the subject thusly:
```ruby
  describe "#wibbles" do
    subject { foo.wibbles }
    it { should have(3).items }
  end
```
Now passing examples will work properly, but failing examples will result in a message like this:
`expected 3 supports_value_expectations?, got 6`

The most common use of the `have` matcher in my experience is testing validations in Rails:
```ruby
  describe "validations" do
    subject { wibble }

    context "with an empty name" do
      before { wibble.name = "" }
      it { should have(1).error_on(:name) }
    end
  end
```

### Your environment
* Ruby version: 3.0.3
* rspec-expectations version: 3.11.0

### Steps to reproduce
```ruby
class Foo
  def wibbles
    [1, 2, 3]
  end
end

RSpec.describe Foo do
  let(:foo) { Foo.new }

  describe "#wibbles" do
    subject { foo }
    it { should have(3).wibbles }
    it { should have_exactly(3).wibbles }
  end
end
```

### Expected behavior
For examples that should pass I would expect passing results.  For examples that should fail I would expect an accurate failure message.

### Actual behavior
In the first example given above, any examples using the `have` matcher throw an exception.  In the second example, failing examples give an inaccurate failure message.

**Comment by JonRowe** [comment]
> the `have` matcher (from rspec-collection_matchers) 

Transferred to rspec-collection_matchers, its likely this just needs to implement `supports_value_expectations?`, want to open a PR for that?

**Comment by amilligan** [comment]
Apologies, I've been traveling.

Yes, I'm happy to start contributing again if you'd like.  I'll take a look in the next day or so.

**Comment by rmeritz** [comment]
Since, I've upgraded to rails-7 and upgraded to `rspec-rails-6.0.0.rc1`, I'm having the same issue. 

Most of my tests are now failing with `NoMethodError:                                                                                                                             undefined method `supports_value_expectations?' for #<SomeActiveModelObject:0x0000556b6f023250>`

**Comment by ekohl** [comment]
@pirj if you now check out the git repo and run `bundle install` (right even that fails), then `bundle exec cucumber` you see that the tests fail.

**Comment by ekohl** [comment]
To provide a bit more context: I wanted to update the [RPM packaging of puppet-lint](https://src.fedoraproject.org/rpms/rubygem-puppet-lint) but that depends on [rspec-collection_matchers](https://src.fedoraproject.org/rpms/rubygem-rspec-collection_matchers). There the test suite is ran on build, but currently failing.

However, I have instead opened a PR to drop the dependency: https://github.com/puppetlabs/puppet-lint/pull/149. That will be easier.

**Comment by ekohl** [comment]
https://github.com/rspec/rspec-collection_matchers/issues/51 is the full report of the failing tests.

**Comment by JonRowe** [comment]
Fixed and released in 1.2.1
